lib/statsd/instrument/datagram.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+class StatsD::Instrument::Datagram < ::String
+  class << self
+    def counter(name, value, type, sample_rate: nil, tags: nil)
+      generic(name, value, -'c', sample_rate: sample_rate, tags: tags)
+    end
+
+
+    # Utility function to convert tags to the canonical form.
+    #
+    # - Tags specified as key value pairs will be converted into an array
+    # - Tags are normalized to only use word characters and underscores.
+    #
+    # @param tags [Array<String>, Hash<String, String>, nil] Tags specified in any form.
+    # @return [Array<String>, nil] the list of tags in canonical form.
+    def normalize_tags(tags)
+      return [] unless tags
+      tags = tags.map { |k, v| k.to_s + ":" + v.to_s } if tags.is_a?(Hash)
+      tags.map { |tag| tag.tr('|,', '') }
+    end
+
+    protected
+
+    def generic(name, value, type, sample_rate: nil, tags: nil)
+      datagram = new
```

Comment:
I don't quite get why it need to be a String subclass. I get this is not finished yet, but at this stage it could perfectly return a regular string.

And if you need to parse a datagram, IMO it's better to do it explicitly by invoking a parser. It's a bit of a slipery slope argument, but I can see some code invoking `#sample_rate` or similar without realizing that it's parsing the string and generating almost 10 objects.

Reply:
> And if you need to parse a datagram

Parsing is basically never needed, *except* for testing: e.g. if you want to verify the right StatsD metric is being emitted. For that use case I am OK taking a bit of a performance hit.

We could use strings and have a separate parser class to make this more explicit. However, with this approach I think it is possible to make this object duck-type compatible with the old Metric class, which will make it a lot easier to migrate existing test suites.

Reply:
Ok, I get it now, but then I'd simply wrap the datagram in such class when exposing it to test helpers.

Reply:
I have moved to using strings.